### DISCUSSION TITLE
destroy() does not have access to the current scope.

### DISCUSSION THREAD

**Discussion by bep** [alpinejs/alpine#2103] [7. Bugs (archived)]
So, custom `data` components can have a `destroy()` function, which is great.

I would expect to use this for _cleaning_, e.g. helping the Garbage Collector do its job. But the `this` does not seem to be the same as the `data` `this`.

**Comment by SimoTod** [comment]
It's the same `this` but `destroy` is the very last call after all the other effects have been deleted and the scope removed from the node so you won't have access to data: https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-data.js#L37
Not sure if it's intentional or just happened to be.

**Reply by bep** [reply]
Ah, ok, thanks -- then I'm not sure what's the use of that would be. This not mentioned in the docus, but there is this test:

```js
destroy() {
   document.querySelector('span').textContent = 'foo'
{
```

Which does not look particulary useful ...

**Reply by SimoTod** [reply]
It would make sense to me if it was inverted (so it would mirror init). I think that it just happened to be like that because it's a new feature without a use case but there are no reasons why it can't be change

**Reply by walirazzaq** [reply]
Came here to open discussion for this very issue, and found it's already there.
@SimoTod I see that `undo()` function returned from `addScopeToNode()` filters data and only parent scopes are available, as you said not sure if it's intended but I haven't found any reference to that.
`addScopeToNode()` is referenced in two other places, in `x-for` and `x-if` but both of them doesn't utilize `undo()` callback. 
In case of nested components we do have access to parent scope

https://github.com/alpinejs/alpine/blob/3c578e5c8df402a49c8d84377b586380231d0500/packages/alpinejs/src/scope.js#L10

**Reply by SimoTod** [reply]
@walirazzaq you don't have to convince me. 😂 Send a PR to change the behaviour, explaining the change and, if there are no reasons not to do it, Caleb will merge the PR. I personally don't see any problems but I'm not the project owner/maintainer.

**Reply by walirazzaq** [reply]
Hahaha tbh, I am a bit afraid of PRs 😆. You got me right though I was literally convincing you as I found you active in alpine repo 😛.
Let's see if I find some more use cases and will try to send a PR

**Reply by SimoTod** [reply]
> Hahaha tbh, I am a bit afraid of PRs 😆. You got me right though I was literally convincing you as I found you active in alpine repo 😛.
> Let's see if I find some more use cases and will try to send a PR  

Haha no worries. This one is quite small tbf but i don't want to send too many PRs at once and I've already a few things I'm looking at in the backlog. 

Also, I'd like more people to be involved so you are more then welcome to give it a go.

Personally, I'm with you. You should be able to see the component scope when calling destroy otherwise it's not really useful.

A use case: if you use setInterval and assign the reference to a component variable, you won't be able to call clearInterval during destroy.

**Reply by walirazzaq** [reply]
I hear you man, you are contributing really helpful stuff ♥️
Yeah, I have something similar use case for event listeners. I'll try to send PR, I have also faced an issue with `click.away` once I have got around a failing test for that I'll send PRs for both 👍🏼

**Comment by adamlogic** [comment]
FYI for anyone who comes here needing this functionality, here's a custom directive that's working for me:

**Edit: Nope, this only seemed like it was working.**


```js
Alpine.directive('cleanup', (_el, { expression }, { evaluate, cleanup }) => {
  cleanup(() => {
    evaluate(expression)
  })
})
```

**Reply by hirasso** [reply]
This directive also can't access the scope, just like `destroy()`

**Reply by SimoTod** [reply]
That won't work because the cleanup callback in x-data will trigger first, removing the component scope.
The real fix is to move undo https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-data.js#L37 after the destroy callback.

**Comment by hirasso** [comment]
Created an issue for this: https://github.com/alpinejs/alpine/issues/2693

**Reply by SimoTod** [reply]
The fix is now on master and it will available in the next release (I created the PR because I don't think anyone was working on it).
Can you close the issue when you have time? This project doesn't really use them. Thanks

**Reply by hirasso** [reply]
@SimoTod that's great to hear! Thanks for creating the PR.
